# ASET: <ins>A</ins>llele <ins>S</ins>pecific <ins>E</ins>xpression <ins>T</ins>oolkit

This nextflow pipeline processes Illumina RNA-Seq reads and generates a data frame that includes allele specific read count data at each SNP site across the samples, annotation for genes and exons. The output data can be taken by the [ASEplot](https://github.com/weishwu/ASEplot) R package for figure generation and genetic ASE test.

## Main steps in ASET:

- RNA-Seq reads are trimmed using `Trimmomatic`.
- One of four alignment approaches can be chosen:
  - `STAR_WASP`: reads are aligned using `STAR` with `--waspOutputMode SAMtag` and a coupled VCF containing SNPs to activate WASP filtering. The alignments that have the `vW` value > 1 are removed.
  - `STAR_NMASK`: reads are aligned using `STAR` against a genome with SNP positions masked as N.
  - `GSNAP`: reads are aligned using `GSNAP` in the SNP-tolerant mode.
  - `ASElux`: reads are aligned and counted using `ASElux`. Currently, untrimmed raw reads are used for ASElux as using trimmed reads causes errors.
- Alignments are filtered, deduplicated, and strand split, unless using `ASElux` which integrates alignment and counting.
- Unless `ASElux` is chosen as the mapper, `ASEReadCounter` from `GATK` is used to calculate allele specific read counts from each strand separately, using the coupled VCF.
- Exons and genes are extracted from the provided GENCODE GTF. Union exons are generated by merging overlapped exons for each gene. ASE sites are then labelled with the gene IDs and union exon coordinates.
- Sample-level contamination can be estimated by calculating non-alt allele frequency at 1/1 sites, and non-ref allele frequency at 0/0 sites (when reference calls are provided in a separate VCF).
- Gene-level maternal contamination for placental samples can be estimated by calculating the mean non-ref allele frequency at 0/0 sites when the maternal genotypes are not 0/0.
- Outputs include trimmed reads, alignments, QC summary (in CSV and MultiQC html), an ASE data frame in TXT, and an Rds file containing the ASE data  and the merged exon coordinates. This file can be taken by the ASEplot R package for figure generation and genetic ASE test.

## Flow charts
- Overview
![](./test_data/ASEprep.drawio.png)
- [DAG if using `STAR_WASP`](./test_data/dag.STAR_WASP.png)
- [DAG if using `STAR_NMASK`](./test_data/dag.STAR_NMASK.png)
- [DAG if using `GSNAP`](./test_data/dag.GSNAP.png)
- [DAG if using `ASElux`](./test_data/dag.ASElux.png)

## Inputs and parameters:
- A sample sheet:
  - For the `from_fastq` mode: ([example](./test_data/sample_sheet.from_fastq.csv))
    - `sample`: sample name
    - `read1`: path to read 1
    - `read2`: path to read 2
    - `snps`: required VCF that contains the heterozygous and homozygous SNPs used for read mapping and ASE read counting
    - `snps_with_ref` (optional): VCF that contains called referece sites, in addition to the variants. If this VCF is provided, it will be used for ASE read counting so that the referece sites will be also examined to check potential contamination ([example](./test_data/sample_sheet.from_fastq.more_snps.csv))
  - For the `from_bam` mode: ([example](./test_data/sample_sheet.from_bam.csv))
    - `sample`: sample name
    - `bam`: path to the alignment file
    - `snps` and `snps_with_ref`: same as above
- params.yaml: ([example](./test_data/params.test.yaml)) 
  - `routine`: can be either `from_fastq` or `from_bam`
  - `mapper`: can be either `STAR_WASP`, `STAR_NMASK`, or `GSNAP`
  - `sample_sheet`: path to the sample sheet
  - `results_dir`: path to the folder where some of the key result files will be copied to
  - `genome_fa`: genome fasta
  - `gtf`: GTF used for gene annotation
  - `rRNA_tRNA_bed`: bed file containing rRNA and tRNA regions, used for Picard CollectRnaSeqMetrics
  - `adapters`: adapter sequences used by Trimmomatic for adapter trimming
  - `rna_strandedness`: either `SECOND_READ_TRANSCRIPTION_STRAND` or `FIRST_READ_TRANSCRIPTION_STRAND`
  - `rna_readlen`: read length, used for STAR genome generation
  - `ase_depth_min`: minimum total read depth required for allele frequency calculation to measure contamination
  - `phase_info` (optional): 
    - `sample`: sample name
    - `phase_info`: phased genotypes to append to the ASE data frame, if available
  - `maternal_vcf` (optional):
    - `sample`: sample name
    - `mat_snps`: maternal genotyping VCF, used for maternal contamination filtration
  - `tool_parameters`: parameters for trimming, alignment, filtering, and ASE read counting. Remember to adjust the `--min-mapping-quality` parameter for ASEReadCounter according to different aligners.

## Output files

- run_report: contains `report.html`, `timeline.html`, `trace.txt`, and `dag.png`.
- results
  - ase_data
    - ASE_hetSNP.txt: ASE read counts on 0/1 SNPs
    - ASE_homSNP.txt: ASE read counts on 1/1 SNPs
    - ASE_homRef.txt: ASE read counts on 0/0 sites
    - ASE_on_hetSNPs.txt.gz: final ASE data frame. Column description is shown in [column_description.txt](./test_data/column_description.txt)
    - ase_data.Rds: final ASE data frame and merged exon coordinates in Rds, which can be taken by the [ASEplot](https://github.com/weishwu/ASEplot) R package for figure generation and genetic ASE test
  - aln
    - 01.raw_aln: alignments from aligner 
    - 02.aln_hc: alignments after filtering
    - 03.aln_hc_dedup: alignments after deduplication
    - 04.aln_hc_dedup_strandsplit: alignments split into two strands
  - qc
    - trimmed_reads: reads trimmed by Trimmomatic
    - fastqc.raw_reads: fastqc report for raw reads
    - fastqc.trimmed_reads: fastqc report for trimmed reads
    - aln_hc_metrics: output from Picard CollectRnaSeqMetrics
    - pre_trim.multiqc.html: MultiQC report summarizing FastQC reports for raw reads
    - post_trim.multiqc.html: MultiQC report summarizing FastQC reports for trimmed reads, Trimmomatic trimming stats, STAR alignment stats, MarkDuplicates deduplication stats, and CollectRnaSeqMetrics output
    - QC_summary.txt: summary of QC metrics in a data frame
  - ref_data
    - gtf_flat.txt: provided GTF converted to a data frame with selected fields
    - ref_gtf_genes.bed: gene coordinates extracted from GTF
    - ref_gtf_exons.bed: exon coordinates extracted from GTF
    - gtf.exonMergedByGenes.bed: union exons merged for each gene
  - tool_versions.txt: versions of the tools used in the pipeline

## Configuration
- ASET comes with four configuration files under the config directory:
  - `env.config`: contains configuration for cluster environment and docker/singularity. Add your own profile according to your environment, or you can simply edit the `cluster` profile if using an HPC cluster.
  - `containers.config`: contains the paths to containers used for running ASET. Normally, initiating ASET will trigger automatic pulling of these containers. Sometimes nextflow can raise errors in pulling. If this happens, try cleaning the singularity cache first by `singularity cache clean -f`, or refer to [this documentation](https://bioinfo-guidelines.readthedocs.io/en/latest/nextflow/trubleshooting.html#failed-to-pull-singularity-image) for more solutions. 
  - `analysis.config`: contains the CPU and memory settings when analyzing real data. Adjust these settings according to your own data. The default settings should work properly for medium-size human RNA-Seq data.
  - `test.config`: contains the resource settings to run a light test run. This test uses 2 CPUs and 2GB memory by default.
- See more details about nextflow configuration at: https://www.nextflow.io/docs/latest/config.html

## Execution instruction

- Download ASET:
```
git clone https://github.com/weishwu/ASET.git
cd ASET
```
- Make sure your environment has nextflow, either singularity or docker, and sufficient computing resource. Nextflow can be installed via conda: https://anaconda.org/bioconda/nextflow
- Edit `config/env.config` to match your computing environment.
- A test run can be initiated by:
```
export NXF_SINGULARITY_CACHEDIR=~/sifs/
export TMPDIR=~/tmp/

nextflow run main.nf \
         -c config/test.config \
         -profile singularity,cluster \
         -params-file test_data/params.test.yaml \
         -with-trace test_data/reports_test/run_report/trace.txt \
         -with-report test_data/reports_test/run_report/report.html \
         -with-timeline test_data/reports_test/run_report/timeline.html \
```
- This test run will pull containers and save to singularity image files under `~/sifs/`. You can change this to a different path. Pulling may take a while. If pulling fails, try cleaning the singularity cache by `singularity cache clean -f`. Specifying the `TMPDIR` directory can avoid saturating the default `/tmp` folder.
- The test run should finish between 10-20 minutes. The output will be in `test_data/reports_test/`.
- To analyze real data, prepare your own sample sheet and params.yaml file, adjust resource settings in `config/analysis.config` if needed, and intiate the run with it. If you have executed the test run and use the same `NXF_SINGULARITY_CACHEDIR`, nextflow will not pull the containers again. Below is an example run:
```
export NXF_SINGULARITY_CACHEDIR=~/sifs/
export TMPDIR=~/tmp/

nextflow run main.nf \
         -c config/analysis.config \
         -profile singularity,cluster \
         -params-file test_data/params.starWasp.yaml \
         -with-trace test_data/reports_starWasp/run_report/trace.txt \
         -with-report test_data/reports_starWasp/run_report/report.html \
         -with-timeline test_data/reports_starWasp/run_report/timeline.html \
```

## Citation
- Manuscript under peer review: https://doi.org/10.21203/rs.3.rs-6844336/v1
